### PR TITLE
tests: Stop and Wait for workspace at the end of each tests

### PR DESCRIPTION
**Pull request by utam0k**
## Description

 Stop and Wait for workspace at the end of each tests to be stabilization

## Related Issue(s)

Relates https://github.com/gitpod-io/gitpod/issues/12248

## How to test

Run the test

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-preview

### CODE REVIEW

**Review by kylos101 — COMMENTED** [review]
What is the total time it takes to run tests assuming no failures? I ask because with increased timeouts, we want to make sure the test run doesn't take too long.

FYI, @jenting shared a test run that had failed tests:
https://werft.gitpod-dev.com/job/gitpod-custom-to-stablization-test.18#integration%20test:test-workspace

Also, I just ran a separate test run in werft, but had trouble with the environment, and notified the platform team:
https://gitpod.slack.com/archives/C03E52788SU/p1662051454479299

I added some questions to the PR, thanks a lot for putting this together, @utam0k !

File: .werft/workspace-run-integration-tests.sh [hunk 1]
```diff
@@ -137,6 +137,8 @@ args=()
 args+=( "-kubeconfig=/home/gitpod/.kube/config" )
 args+=( "-namespace=default" )
 [[ "$USERNAME" != "" ]] && args+=( "-username=$USERNAME" )
+args+=( "-timeout=60m" )
```
[the remark this answers is not in the document]
Comment (utam0k) [reply .werft/workspace-run-integration-tests.sh:140]: Yes, it is, because these args pass each test component.

**Comment by utam0k** [comment]
> FYI https://werft.gitpod-dev.com/job/gitpod-custom-to-stablization-test.18

@jenting @kylos101 
This PR improves stabilization, not complete fixing the integration test. I have already found the below issues
- port-forwarding sometimes fails
- sometimes hit https://github.com/gitpod-io/gitpod/issues/12021 when trying to stop a workspace

File: test/pkg/integration/integration.go [hunk 2]
```diff
@@ -187,20 +198,21 @@ func Instrument(component ComponentType, agentName string, namespace string, kub
...
+	}
+	go execF()
 	select {
 	case err := <-execErrs:
 		if err != nil {
 			return nil, closer, err
 		}
 		return nil, closer, fmt.Errorf("agent stopped unexepectedly")
-	case <-time.After(1 * time.Second):
+	case <-time.After(30 * time.Second):
```
[the remark this answers is not in the document]
Comment (utam0k) [reply test/pkg/integration/integration.go:215]: I accused this time means waiting time for the start-up of the agent

**Comment by jenting** [comment]
> > FYI https://werft.gitpod-dev.com/job/gitpod-custom-to-stablization-test.18
> 
> @jenting @kylos101 This PR improves stabilization, not complete fixing the integration test. I have already found the below issues
> 
> * port-forwarding sometimes fails [The flakiness of workspace integration test when port-forwarding #10671](https://github.com/gitpod-io/gitpod/issues/10671)
> * sometimes hit [[ws-manager] The container could not be located when the pod was terminated #12021](https://github.com/gitpod-io/gitpod/issues/12021) when trying to stop a workspace

@utam0k I add the port-forwarding issue link 😉

**Comment by utam0k** [comment]
> > > FYI https://werft.gitpod-dev.com/job/gitpod-custom-to-stablization-test.18
> > 
> > 
> > @jenting @kylos101 This PR improves stabilization, not complete fixing the integration test. I have already found the below issues
> > 
> > * port-forwarding sometimes fails [The flakiness of workspace integration test when port-forwarding #10671](https://github.com/gitpod-io/gitpod/issues/10671)
> > * sometimes hit [[ws-manager] The container could not be located when the pod was terminated #12021](https://github.com/gitpod-io/gitpod/issues/12021) when trying to stop a workspace
> 
> @utam0k I add the port-forwarding issue link 😉

Oh, I missed it. But I'm trying to fix it on https://github.com/gitpod-io/gitpod/issues/12248 so that I will close your issue. Thanks

**Comment by jenting** [comment]
Thanks for your effort @utam0k 

I thought all the integration test cases failed for the same reason https://github.com/gitpod-io/gitpod/issues/10671
But with this PR introduced changes, it seems no. And I summarize two reasons for the flaky integration test
- we can't have multiple workspaces when running the integration test (I think we could support it, no?)
- the timeout issue (probably due to preview env CPU/Memory/Disk are slow...)

**Comment by utam0k** [comment]
> Thanks for your effort @utam0k
> 
> I thought all the integration test cases failed for the same reason #10671 But with this PR introduced changes, it seems no. And I summarize two reasons for the flaky integration test
> 
> * we can't have multiple workspaces when running the integration test (I think we could support it, no?)
> * the timeout issue (probably due to preview env CPU/Memory/Disk are slow...)

Yes, I thought both were the same reason, lack of resources. Ideally, we can run all integration tests in parallel, but first I think we should focus on passing the integration; next, we can try it.

**Review by jenting — APPROVED** [review]
LGTM

Let's see how it improves the integration test robustness 💪

**Comment by kylos101** [comment]
> > Thanks for your effort @utam0k
> > I thought all the integration test cases failed for the same reason #10671 But with this PR introduced changes, it seems no. And I summarize two reasons for the flaky integration test
> > 
> > * we can't have multiple workspaces when running the integration test (I think we could support it, no?)
> > * the timeout issue (probably due to preview env CPU/Memory/Disk are slow...)
> 
> Yes, I thought both were the same reason, lack of resources. Ideally, we can run all integration tests in parallel, but first I think we should focus on passing the integration; next, we can try it.

Indeed, @utam0k , great point. In other words, get tests to pass (so we have confidence), then make test runs faster (so we can have shorter feedback loops). [Some related thoughts.](https://wiki.c2.com/?MakeItWorkMakeItRightMakeItFast)

**Pull request merged**